ARCHITECTURE.md

```diff
@@ -0,0 +1,183 @@
+This document tries to describe the organising principles behind the
+overlap checker.
+
+These tools are inspired by the https://github.com/ukaea/parallel-preprocessor/
+previously written by Qingfeng Xia.
+
+# Users
+
+There are currently two groups of users these tools are being designed
+to support:
+
+1. CAD-CI, i.e. automatic validation of CAD models. Currently this is
+   just making sure that models are self consistent, no significant
+   overlaps.
+2. Neutronics modelling, i.e. preprocessing CAD models so they can be
+   used with tools like DAGMC.
+
+# Terminology
+
+ * A **solid**s is a 3D volume described by a set of faces, edges, and
+   verticies.
+ * Solids **intersect** when some point exists that is within either
+   solids or on the surface of either solid. It's useful to
+   distinguish between these cases:
+   * Intersecting solids **touch** when the volume of intersection is
+     empty. For example, a vertex of one solid lies on the face of
+     another. Or, minimally, when two vertices are at the same
+     location.
+   * Intersection solids **overlap** when the volume of the
+     intersection is not empty. E.g. points exist that are within the
+     shapes.
+ * Solids are **distinct** when they do not intersect in any way. For
+   example, their bounding boxes do not overlap. Note that this can
+   get awkward to determine, e.g. a cylinder could have a torus around
+   it whose "hole" is larger diameter than the cylinder.
+
+# Model representation
+
+The parts of the CAD we really care about are the solid parts and
+their physical properties, like material. Some operations mutate these
+parts (e.g. the imprint step) and we still want to be able to
+reference the parts consistently.
+
+CAD models tend to be hierarchical (it makes it easier to work on them
+like this) but this makes referring to parts awkward. The way I've
+chosen is to "linearise" the hierarchy of solids into a list, and
+maintain this linear order across related files and operations.
```

Comment:
A key point here is that you're not creating any new solids (e.g. with solid1, solid2, and intersection12), but are adding the intersection to one of the original solids.

Reply:
Another question is whether you can get the solids out of the STEP file in a deterministic order?

Reply:
getting files out of a STEP file in deterministic order shouldn't matter, that's just done once when converting to BREP.  it's whether running these different utilities iterate over the BREP solids they get them back in the same order

Reply:
> A key point here

this was what the following paragraph (i.e. "For example, when checking") is giving examples of?